contracts/external/nexus/WrappedCoveredPrincipalToken.sol

```diff
@@ -124,6 +153,36 @@ contract WrappedCoveredPrincipalToken is ERC20PermitWithSupply, Authorizable {
         _mint(msg.sender, _amount);
     }
 
+    /// @notice Tell whether the given `_wrappedPosition` is whitelisted or not.
+    /// @param  _wrappedPosition Address of the wrapped position.
+    /// @return returns boolean, True -> allowed otherwise false.
+    function isAllowedWp(address _wrappedPosition) public view returns (bool) {
+        return _allowedWrappedPositions.contains(_wrappedPosition);
+    }
+
+    /// @notice Returns the list of wrapped positions that are whitelisted with the contract.
+    ///         Order is not maintained.
+    /// @return Array of addresses.
+    function allWrappedPositions() public view returns (address[] memory) {
+        return _allowedWrappedPositions.values();
+    }
+
+    /// @notice Reclaim tranche token (i.e principal token) by the authorized account.
+    /// @param  _expiration Timestamp at which the derived tranche would get expired.
+    /// @param  _wrappedPosition Address of the Wrapped position which is used to derive the tranche.
+    function reclaimPt(uint256 _expiration, address _wrappedPosition)
```

Comment:
I think this function should burn sender's balance and send them the corresponding amount of underlying. As it stands, the `RECLAIM_ROLE` can just rug the whole underlying at any moment.

I expect `RECLAIM_ROLE` to be given to a third party like Nexus mutual so I don't want this trust in the system.

Reply:
I don't think this makes sense. The wrapping is a one way function so you can't really 'rug' the backing token because a wrappers can't get any funds out anyway. Also we don't want to require Nexus to have the tokens because we want to give them max flexibility to redeem even if their architecture means a sc without generic call holds the tokens.

Reply:
> The wrapping is a one way function so you can't really 'rug' the backing token because a wrappers can't get any funds out anyway.

This function sends the tranche tokens to the user. aka it is unwrapping the wrapped tokens without consumed the wrapped tokens.